### PR TITLE
fix(boot.go,jobs/jobs.go): fix panic on startup

### LINKED ISSUE #54: workflow manager panics on startup
```console
$ kd logs -f deis-workflow-manager-emz6z
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x651b0d]

goroutine 12 [running]:
panic(0x12456e0, 0xc82000a070)
    /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/deis/workflow-manager/data.getDeisRCItems(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/deis/workflow-manager/data/data.go:153 +0x14d
github.com/deis/workflow-manager/data.InstalledDeisData.Get(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/deis/workflow-manager/data/installed_data.go:29 +0x6b
github.com/deis/workflow-manager/data.(*InstalledDeisData).Get(0xc820248530, 0x0, 0x0, 0x0, 0x0, 0x0)
    <autogenerated>:16 +0xb5
github.com/deis/workflow-manager/data.GetInstalled(0x7faf07d3c6f0, 0xc820248530, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/deis/workflow-manager/data/data.go:86 +0x75
github.com/deis/workflow-manager/data.GetCluster(0x7faf07d3c6f0, 0xc820248530, 0x7faf07d3b4d8, 0xc82025a030, 0x7faf07d3b510, 0xc820246f80, 0x0, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/deis/workflow-manager/data/data.go:31 +0x84
github.com/deis/workflow-manager/data.NewAvailableVersionsFromAPI.func1(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/deis/workflow-manager/data/available_versions.go:49 +0x35c
github.com/deis/workflow-manager/data.availableVersionsFromAPI.Refresh(0x0, 0x0, 0x0, 0xc820246f40, 0xc82025a000, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/deis/workflow-manager/data/available_versions.go:56 +0x5b
github.com/deis/workflow-manager/data.(*availableVersionsFromAPI).Refresh(0xc82015b680, 0x0, 0x0, 0x0, 0x0, 0x0)
    <autogenerated>:4 +0xbe
github.com/deis/workflow-manager/jobs.GetLatestVersionData.Do(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/deis/workflow-manager/jobs/jobs.go:58 +0x84
github.com/deis/workflow-manager/jobs.(*GetLatestVersionData).Do(0xc820246ca0, 0x0, 0x0)
    <autogenerated>:2 +0xa2
github.com/deis/workflow-manager/jobs.runJobs(0xc820246c80, 0x2, 0x2)
    /go/src/github.com/deis/workflow-manager/jobs/jobs.go:89 +0xbb
github.com/deis/workflow-manager/jobs.DoPeriodic.func1(0xc820246c80, 0x2, 0x2, 0xc82015b380, 0xc8201e7800)
    /go/src/github.com/deis/workflow-manager/jobs/jobs.go:72 +0x4a
created by github.com/deis/workflow-manager/jobs.DoPeriodic
    /go/src/github.com/deis/workflow-manager/jobs/jobs.go:82 +0x99
```

### PR DESCRIPTION
**Pull request by arschles**
# Summary Of Issues

I opened this pull request to fix the panic described in #54. After fixing the panic described in that issue, I uncovered the following other bugs, in order of the below listing. Since each bug uncovers the next in turn, I believe that the fixes all need to be included into a single PR, even if this PR is large in scope.

- The code was passing an incorrect `time.Duration` to the poller
  - It was effectively passing `time.Duration(config.Spec.Polling)*time.Second*time.Second` into `time.NewTicker`, which was a negative number and caused `time.NewTicker` to return an error
- The first recurring job is infinitely recurring (see #55, TODO below)
- The `VERSIONS_API_URL` and `DOCTOR_API_URL` environment variables (among others) are not being set properly in the deis chart
  - PR forthcoming for that issue. For now, see https://github.com/deis/charts/issues/267 (fixed by https://github.com/deis/charts/pull/268) for tracking

# Summary of Fixes

This PR include the following fixes to address the above issues:

- Creates a new `rest` package that has a `Client` interface, whose implementation carries a shared `*http.Client` and the base URL for the workflow-manager API server. As such, that data need not be passed everywhere the client is used
- Specifies defaults in the `envconfig` struct for the `VERSIONS_API_URL`, `DOCTOR_API_URL`, and `API_VERSION` environment variables. These defaults allow this component to function until https://github.com/deis/charts/issues/267 is fixed
- Refactors the interface implementations in the `data` package to be non-exported. This change forces other packages to create these implementations via "constructor" functions, which ensures that all required fields are given. This design pattern forces all other packages to create new implementations in a single place, which greatly reduces the chances that other packages initialize an implementation with `nil` members, and fixes #54 
- Fixes the infinite recursion specified in #55 by passing a `types.Cluster` to `(github.com/deis/workflow-manager/data).AvailableVersions`'s `Refresh` func. With this change, `AvailableVersions` no longer needs to call `GetCluster`, which was the source of the recursion
- Changes various implementations in the `data` package to no longer swallow errors. All code in `data` should now check for and return errors
- Similar to the above, code in the `data` package no longer logs errors. Instead, implementations just return the errors and rely on callers to handle logging, handling, etc...
- Passing a `time.Duration` to `jobs.DoPeriodic` and not multiplying that `time.Duration` by `time.Second`. Previously, the resultant `time.Duration` was an overflow and `time.NewTimer` was returning an error
- Creating tests for `jobs/jobs.go`

Fixes https://github.com/deis/workflow-manager/issues/55
Fixes https://github.com/deis/workflow-manager/issues/54

Still TODO

- [x] fix throttled reqs (see https://github.com/deis/workflow-manager/issues/55 for what logs look like)
- [x] fix tests

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -6,9 +6,9 @@ import "github.com/kelseyhightower/envconfig"
 type Specification struct {
 	Port           string `default:"8080" envconfig:"PORT"`
 	Polling        int    `default:"43200" envconfig:"POLL_INTERVAL_SEC"` // 43200 seconds = 12 hours
-	VersionsAPIURL string `envconfig:"VERSIONS_API_URL"`
-	DoctorAPIURL   string `envconfig:"DOCTOR_API_URL"`
-	APIVersion     string `envconfig:"API_VERSION"`
+	VersionsAPIURL string `envconfig:"VERSIONS_API_URL" default:"https://versions-staging.deis.com"`
```
Comment (bacongobbler) [inline config/config.go]: :+1: I just documented this as the defaults, but felt it was odd that the chart defined these envvars

File: data/cluster_id.go [hunk 2]
```diff
@@ -53,16 +51,7 @@ func NewClusterIDFromPersistentStorage(sgc KubeSecretGetterCreator) ClusterID {
...
-		switch e := err.(type) {
-		case *k8sErrors.StatusError:
-			// If the error isn't a 404, we don't know how to deal with it
-			if e.ErrStatus.Code != 404 {
-				return "", err
-			}
-		default:
-			return "", err
-		}
+		return "", err
```
Comment (bacongobbler) [inline data/cluster_id.go]: so the idea now is that if k8s returns a 404, it's still an error from `Get()`'s perspective?

**Comment by bacongobbler** [comment]
From my naive surface-deep code review (not too well versed with workflow-manager... Need to do a deep dive one day), this all looks good to me.

File: data/cluster_id.go (hunk 2 above)
Comment (arschles) [inline data/cluster_id.go]: @bacongobbler yes. I didn't want to swallow the error if Kubernetes returned a different error code, as it seemed to do when rate limiting a request

**Pull request merged**